kiwi_keg/generator.py

```diff
@@ -153,6 +156,94 @@ def create_custom_scripts(self, override: bool = False):
+                                overlay_path
+                            )
+                            # loop all the file paths of overlay sub directory 'overlay_path'
+                            for name in KegUtils.get_all_files(overlay_full_path):
+                                rel_path = os.path.relpath(name, overlay_full_path)
+                                new_dir = os.path.dirname(rel_path)
+                                if new_dir:
+                                    new_dir = os.path.join(overlay_dest_dir, new_dir)
+                                    os.makedirs(new_dir, exist_ok=True)
+                                dest_file = os.path.join(overlay_dest_dir, rel_path)
+                                shutil.copy(name, dest_file)
+
+                        tarball_data[overlay_name] = {}
+                        tarball_data[overlay_name] = overlay_dest_dir
+
+                    if not disable_tar_overlays:
+                        for overlay_name, overlay_dest_dir in tarball_data.items():
+                            overlay_tarball_name = '{}.tar.gz'.format(overlay_name)
+                            self._create_tarball(
+                                os.path.join(self.dest_dir, overlay_tarball_name),
+                                overlay_dest_dir
+                            )
+                            shutil.rmtree(overlay_dest_dir)
+                            self._update_config_kiwi(
+                                overlay_tarball_name,
+                                overlay_dest_dir
+                            )
+
+        if not disable_tar_overlays and not has_overlays:
+            log.warn(
+                'Attempt to create a tarball but not overlay paths were provided.'
+            )
+
+    @staticmethod
+    def _create_tarball(tarball_dir, dest_dir):
+        with tarfile.open(tarball_dir, 'w:gz') as tar:
+            for overlay_dir in os.scandir(dest_dir):
+                tar.add(
+                    overlay_dir.path,
+                    arcname=overlay_dir.name
+                )
+
+    def _update_config_kiwi(self, archive_name, dest_dir):
+        if 'root' != archive_name.partition('.')[0]:
+            if os.path.exists(dest_dir) and os.path.exists(self.kiwi_description):
+                etree_parser = etree.XMLParser(remove_blank_text=True)
+                kiwi_xml = etree.parse(self.kiwi_description, parser=etree_parser)
+                kiwi_root = kiwi_xml.getroot()
+
+                image_element = kiwi_root.find('.//packages[@type="image"]')
+
+                archive_element = etree.SubElement(image_element, 'archive')
+                archive_element.attrib['name'] = archive_name
+                tree = etree.ElementTree(kiwi_root)
+                tree.write(
+                    self.kiwi_description,
+                    encoding="utf-8",
+                    xml_declaration=True,
+                    pretty_print=True
+                )
```

Comment:
Wouldn't this add all archives simply to the first `<packages type="image">` section? Doesn't really matter though, I still think we shouldn't handle it like that in the first place, i.e. rewrite the just generated `config.kiwi`. All the information is there as soon as keg parsed the input data, so there's no need to do it that way. It's needlessly complicated IMHO and it assumes a certain structure of the `config.kiwi` in keg which is actually defined in the schema and not by keg. As mentioned before, I'd rather preprocess the data in keg and generate the complete `config.kiwi` in one go.

Reply:
True, and agreed, I rather doing that refactor once this is merged, in a new different PR